### PR TITLE
Setup E2E test using NATS Backend

### PR DESCRIPTION
**Pull request by grischperl**
## Description
Add Github Action to run E2E test

**Changes proposed in this pull request:**

- EPP Image from PR is used in Eventing Manager
- Execute E2E Tests from EM
- E2E uses NATS Backend

**Related issue(s)**

[kyma/#18477](https://github.com/kyma-project/kyma/issues/18477)

### CODE REVIEW

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      - main
+      - "release-*"
+    paths-ignore:
+      - "docs/**"
+      - "**.md"
+      - "sec-scanners-config.yaml"
+
+env:
+  NATS_MANAGER_IMAGE: europe-docker.pkg.dev/kyma-project/prod/nats-manager:latest
+  EVENTING_MANAGER_IMAGE: europe-docker.pkg.dev/kyma-project/prod/eventing-manager:latest
```
Comment (mfaizanse) [inline .github/workflows/e2e.yml:15]: Should we use latest image tag? or image from latest release of eventing+nats manager?

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          # The check interval is kept long otherwise it will exhaust the GitHub rate limit (More info: https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#rate-limiting)
+          check_interval: 60000 # 1 minute in milliseconds
+        env:
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
+          GITHUB_OWNER: "${{ github.repository_owner }}"
+          GITHUB_REPO: "eventing-publisher-proxy"
+
+      - name: Setup Eventing Manager with new EPP image
+        run: |
+          kubectl patch deployment eventing-manager -n kyma-system --type='json' -p='[{"op": "replace", "path": "/spec/template/spec/containers/0/env/8/value", "value": "'$EPP_IMAGE'"}]'
```
Comment (mfaizanse) [inline .github/workflows/e2e.yml:96]: So if we later do any changes to ENVs of eventing-manager deployment, will this command fail? as I see the indexes are hard coded `0` and `8`?

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        run: |
+          make -C eventing-manager e2e-setup
+
+      - name: Setup subscriptions for testing
+        run: |
+          make -C eventing-manager e2e-eventing-setup
+
+      - name: Test eventing
+        run: |
+          make -C eventing-manager e2e-eventing
```
Comment (mfaizanse) [inline .github/workflows/e2e.yml:108]: Should we also run the test cleanup step?

File: .github/workflows/e2e.yml (hunk 2 above)
Comment (mfaizanse) [reply .github/workflows/e2e.yml:96]: Maybe we can get the `json` and use `jq` to find the right env with key and replace its value and apply it back?

File: .github/workflows/e2e.yml (hunk 1 above)
Comment (friedrichwilken) [reply .github/workflows/e2e.yml:15]: I was wondering the same. is there a way to always get `latest-rel`?

Comment (grischperl) [reply .github/workflows/e2e.yml:15]: `latest` will ensure the current EPP changes work with the newest changes of EM and NM. If we use a release tag, we need to update the image every time there is a new release. If we prefer a stable environment by using a release tag, I'd try and see if I can refer to these env variables from somewhere else, so we don't have to remember to bump them in the test itself.

File: .github/workflows/e2e.yml (hunk 2 above)
Comment (grischperl) [reply .github/workflows/e2e.yml:96]: Yes. I'll check to improve the patch command to not rely on the indexes.

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (grischperl) [reply .github/workflows/e2e.yml:108]: Yes!

File: .github/workflows/e2e.yml (hunk 2 above)
Comment (grischperl) [reply .github/workflows/e2e.yml]: Done

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (grischperl) [reply .github/workflows/e2e.yml:111]: Done

File: .github/workflows/e2e.yml (hunk 1 above)
Comment (grischperl) [reply .github/workflows/e2e.yml]: I used `curl` to get the latest release for the Eventing Manager. Same can be done for NATS Manager as soon as there is a new release.

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+          kubectl create ns kyma-system
+
+      - name: Setup NATS Manager
+        run: |
+          make -C nats-manager install
+
+      - name: Deploy NATS Manager
+        run: |
+          version=$(curl -i https://github.com/kyma-project/nats-manager/releases/latest | awk -F '/' '/^location/ {print  substr($NF, 1, length($NF)-1)}')
+          make -C nats-manager deploy IMG=$NATS_MANAGER_IMAGE:v20231013-6091000e
```
Comment (grischperl) [inline .github/workflows/e2e.yml:68]: `v20231013-6091000e` is hardcoded until we have a new tagged release of NATS Manager

File: .github/workflows/e2e.yml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
+          GITHUB_OWNER: "${{ github.repository_owner }}"
+          GITHUB_REPO: "eventing-publisher-proxy"
+
+      - name: Setup Eventing Manager with new EPP image
+        run: |
+          kubectl get deployment eventing-manager -n kyma-system -o=json |
+          jq --arg new_image "$EPP_IMAGE" \
+          '.spec.template.spec.containers[0].env |= map(if .name == "PUBLISHER_IMAGE" then .value = $new_image else . end)' |
+          kubectl apply -f -
```
Comment (mfaizanse) [inline .github/workflows/e2e.yml:85]: Should we wait for the deployment to rollout the new env to all Pods? before moving to next step

Comment (grischperl) [reply .github/workflows/e2e.yml:85]: Added `kubectl rollout status`

File: .github/workflows/e2e.yml [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        env:
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
+          GITHUB_OWNER: "${{ github.repository_owner }}"
+          GITHUB_REPO: "eventing-publisher-proxy"
+
+      - name: Setup Eventing Manager with new EPP image
+        run: |
+          kubectl get deployment eventing-manager -n kyma-system -o=json |
+          jq --arg new_image "$EPP_IMAGE" \
+          '.spec.template.spec.containers[0].env |= map(if .name == "PUBLISHER_IMAGE" then .value = $new_image else . end)' |
```
[the remark this answers is not in the document]
Comment (grischperl) [reply .github/workflows/e2e.yml:84]: Added step to compare desired image with image in EPP deployment

**Review by mfaizanse — APPROVED** [review]

**Pull request merged**